src/emit/ops/cfg.rs

```diff
@@ -0,0 +1,288 @@
+            let outputs_rmb = context.node_ins_rmb(o)?;
+            // read the values from our input node
+            let inputs = inputs_rmb.read_vec(context.builder(), [])?;
+
+            // emit all our children and read the values from the rowmailbox of our output node
+            emit_dataflow_parent(
+                context,
+                EmitOpArgs {
+                    node: node.clone(),
+                    inputs,
+                    outputs: outputs_rmb.promise(),
+                },
+            )?;
+            let outputs = outputs_rmb.read_vec(context.builder(), [])?;
+
+            // We create a helper block per-tag. We switch to the helper block,
+            // where we then store the input args for the successor block, then
+            // unconditionally branch.
+            //
+            // We use switch even when we have 1 or 2 successors, where we could
+            // use unconditional branch or conditional branch, to simplify the
+            // code here at the expense of messier generated code. We expect the
+            // simplify-cfg pass to clean this up without issue.
+            let branch_sum_type = SumType::new(node.sum_rows.clone());
+            let llvm_sum_type = context.llvm_sum_type(branch_sum_type)?;
+            let tag_bbs = successor_data
+                .into_iter()
+                .enumerate()
+                .map(|(tag, (target_bb, target_rmb))| {
+                    let bb = context.build_positioned_new_block("", Some(*bb), |context, bb| {
+                        let builder = context.builder();
+                        let mut vals =
+                            llvm_sum_type.build_untag(builder, tag as u32, outputs[0])?;
+                        vals.extend(&outputs[1..]);
+                        target_rmb.write(builder, vals)?;
+                        builder.build_unconditional_branch(target_bb)?;
+                        Ok::<_, anyhow::Error>(bb)
+                    })?;
+                    Ok((
+                        llvm_sum_type.get_tag_type().const_int(tag as u64, false),
+                        bb,
+                    ))
+                })
+                .collect::<Result<Vec<_>>>()?;
+            let tag_v = llvm_sum_type.build_get_tag(context.builder(), outputs[0])?;
+            context
+                .builder()
+                .build_switch(tag_v, tag_bbs[0].1, &tag_bbs[1..])?;
+            Ok(())
+        })
+    }
+}
+
+impl<'c, H: HugrView> EmitOp<'c, ExitBlock, H> for CfgEmitter<'c, '_, H> {
+    fn emit(&mut self, args: EmitOpArgs<'c, ExitBlock, H>) -> Result<()> {
+        let outputs = self.take_outputs()?;
+        let (bb, inputs_rmb) = self.bbs.get(&args.node().generalise()).unwrap();
+        self.context.build_positioned(*bb, |context| {
+            let builder = context.builder();
+            outputs.finish(builder, inputs_rmb.read_vec(builder, [])?)
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Are non-local dom edges also supported? It should probably work just fine with the mailbox system, right? In that case adding an extra test might be good